### PR TITLE
Add parameter skip_under_replicated_partition and skip_reorder_replica to demote_broker endpoint.

### PR DESCRIPTION
**Pull request by kidkun**
patch for issue[403](https://github.com/linkedin/cruise-control/issues/403)

### CODE REVIEW

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/PreferredLeaderElectionGoal.java [hunk 1]
```diff
@@ -38,10 +56,20 @@ public boolean optimize(ClusterModel clusterModel, Set<Goal> optimizedGoals, Set
     Set<TopicPartition> partitionsToMove = new HashSet<>();
     for (Broker b : clusterModel.demotedBrokers()) {
       for (Replica r : b.replicas()) {
-        Partition p = clusterModel.partition(r.topicPartition());
-        p.moveReplicaToEnd(r);
+        // There are two scenarios where replica swap operation is skipped:
+        // 1.the replica on the to-be-demoted broker is not leader replica and skip_reorder_replica parameter is set in request.
+        // 2.the replica on the to-be-demoted broker is leader replica, the partition is currently under replicated and
+        // skip_under_replicated_partition parameter is set in request.
+        if (!(_skipUnderReplicatedPartition && r.isLeader() && isPartitionUnderReplicated(r))
```
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/PreferredLeaderElectionGoal.java]: What if (1) the replica on a demoted broker is currently a follower and (2) `_skipUnderReplicatedPartition` is `true`? IIUC, there will still be a replica swap, in which this replica moves to the end of replica list in the corresponding follower.

Can we also handle this scenario?

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/PreferredLeaderElectionGoal.java [hunk 2]
```diff
@@ -67,6 +95,11 @@ public boolean optimize(ClusterModel clusterModel, Set<Goal> optimizedGoals, Set
     return true;
   }
 
+  private boolean isPartitionUnderReplicated(Replica r) {
```
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/PreferredLeaderElectionGoal.java]: Currently it seems a little difficult to understand which replica to pass to this function.

Would the change of the parameter from `Replica` to `Partition` and passing `r.topicPartition()` to this function make things more clear?

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/PreferredLeaderElectionGoal.java [hunk 3]
```diff
@@ -38,10 +56,20 @@ public boolean optimize(ClusterModel clusterModel, Set<Goal> optimizedGoals, Set
     Set<TopicPartition> partitionsToMove = new HashSet<>();
     for (Broker b : clusterModel.demotedBrokers()) {
       for (Replica r : b.replicas()) {
-        Partition p = clusterModel.partition(r.topicPartition());
-        p.moveReplicaToEnd(r);
+        // There are two scenarios where replica swap operation is skipped:
```
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/PreferredLeaderElectionGoal.java]: We need a unit test to verify this functionality.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/PreferredLeaderElectionGoal.java [hunk 4]
```diff
@@ -38,10 +56,20 @@ public boolean optimize(ClusterModel clusterModel, Set<Goal> optimizedGoals, Set
...
+        // 1.the replica on the to-be-demoted broker is not leader replica and skip_reorder_replica parameter is set in request.
+        // 2.the replica on the to-be-demoted broker is leader replica, the partition is currently under replicated and
```
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/PreferredLeaderElectionGoal.java]: Why does it need to be the leader replica?

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/async/DemoteBrokerRunnable.java [hunk 5]
```diff
@@ -13,14 +13,16 @@
...
+ * com.linkedin.kafka.cruisecontrol.async.progress.OperationProgress, boolean, Integer, boolean, boolean, HttpServletRequest)}
  */
 public class DemoteBrokerRunnable extends OperationRunnable {
   private final Collection<Integer> _brokerIds;
   private final boolean _dryRun;
   private final boolean _allowCapacityEstimation;
   private final Integer _concurrentLeaderMovements;
   private final HttpServletRequest _request;
+  private boolean _skipUnderReplicatedPartition;
+  private boolean _skipReorderReplica;
```
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/async/DemoteBrokerRunnable.java]: Any reason regarding why these variables are not `final`?

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/parameters/ParameterUtils.java [hunk 6]
```diff
@@ -70,6 +70,8 @@
   private static final String SKIP_HARD_GOAL_CHECK_PARAM = "skip_hard_goal_check";
   private static final String EXCLUDED_TOPICS_PARAM = "excluded_topics";
   private static final String USER_TASK_IDS_PARAM = "user_task_ids";
+  private static final String SKIP_UNDER_REPLICATED_PARTITION_PARAM = "skip_under_replicated_partition";
+  private static final String SKIP_REORDER_REPLICA_PARAM = "skip_reorder_replica";
```
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/parameters/ParameterUtils.java]: Please see my comments regarding the name of these parameters.

File: cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/analyzer/PreferredLeaderElectionGoalTest.java [hunk 7]
```diff
@@ -50,7 +50,7 @@
   public void testOptimize() throws KafkaCruiseControlException {
     ClusterModel clusterModel = createClusterModel();
 
-    PreferredLeaderElectionGoal goal = new PreferredLeaderElectionGoal();
+    PreferredLeaderElectionGoal goal = new PreferredLeaderElectionGoal(false, false, null);
```
Comment (efeg) [inline cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/analyzer/PreferredLeaderElectionGoalTest.java]: We are missing unit tests for the new functionality.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/PreferredLeaderElectionGoal.java (hunk 1 above)
Comment (kidkun) [reply cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/PreferredLeaderElectionGoal.java]: Good point. I agree `_skipUrpDemotion`(original `_skipUnderReplicatedPartition`) should cover this scenario.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/PreferredLeaderElectionGoal.java [hunk 8]
```diff
@@ -29,6 +31,21 @@
  */
 public class PreferredLeaderElectionGoal implements Goal {
   private static final Logger LOG = LoggerFactory.getLogger(PreferredLeaderElectionGoal.class);
+  private final boolean _skipUrpDemotion;
+  private final boolean _excludeFollowerDemotion;
+  private final Cluster _kafkaCluster;
+
+  public PreferredLeaderElectionGoal() {
```
[the remark this answers is not in the document]
Comment (kidkun) [reply cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/PreferredLeaderElectionGoal.java]: Some of the reflective instance creation used this constructor, one example is `goalsByPriority()`

**Pull request merged**